src/protocol/codecs/topic.rs

```diff
@@ -0,0 +1,113 @@
+use std::io::{self, ErrorKind};
+
+use bytes::BytesMut;
+use tokio_util::codec::Decoder;
+
+use crate::protocol::invalid_data;
+
+/// Topic is a key-value string pair.
+pub struct Topic {
+    /// Key.
+    pub key: String,
+
+    /// Value.
+    pub value: String,
+}
+
+#[derive(Default, Clone)]
+pub struct TopicCodec;
+
+impl Decoder for TopicCodec {
+    type Item = Vec<Topic>;
+    type Error = io::Error;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        if src.is_empty() {
+            return Ok(Some(vec![]));
+        }
+
+        let (&num_topics, mut src) = src
+            .split_first()
+            .ok_or_else(|| invalid_data!("couldn't decode the number of topics"))?;
+        let mut topics = Vec::with_capacity(num_topics as usize);
+
+        for _ in 0..num_topics {
+            let (&key_len, buf) = src
+                .split_first()
+                .ok_or_else(|| invalid_data!("couldn't decode the key length"))?;
+
+            if key_len as usize > buf.len() {
+                return Err(invalid_data!("invalid topic length"));
+            }
+
+            let (key, buf) = buf.split_at(key_len as usize);
+
+            let (&val_len, buf) = buf
+                .split_first()
+                .ok_or_else(|| invalid_data!("couldn't decode the value length"))?;
+
+            if val_len as usize > buf.len() {
+                return Err(invalid_data!("invalid topic length"));
+            }
+
+            let (val, buf) = buf.split_at(val_len as usize);
+
+            let key = String::from_utf8(key.to_vec()).map_err(|_| ErrorKind::InvalidData)?;
+            let value = String::from_utf8(val.to_vec()).map_err(|_| ErrorKind::InvalidData)?;
+            topics.push(Topic { key, value });
+
+            // prepare the src for the next iteration.
+            src = buf;
```

Comment:
this will probably not be needed if the uses of `split_at` are changed to the aforementioned methods